### PR TITLE
Add: implement DeepSeek V4.1 Flash decode C2A reuse operator

### PR DESCRIPTION
**Pull request by indigo1973**
Add: implement DeepSeek V4.1 Flash decode C2A reuse operator

Implement A5 decode C2A reuse with the shared SWA and C2A Full
projection, normalization, rotary, cache and sparse-attention kernels.
Consume packed FP4 compressed caches and published physical Top-K rows
without a compressor or indexer.

Publish only mapped window rows and respect the active token prefix.
Accumulate WoA in FP32 and retain FP32 output partials through the
shared TP reduction and consecutive-epoch synchronization.

Provide deterministic in-file Golden Harness validation for TP1/2/4
and DP1/2, including masked, zero, padded and replayed inputs.
Use a Full-aligned CPU reference and comparators while retaining the
original public golden function. Check untouched cache bytes and
inactive output rows against their input state.

### CODE REVIEW

File: models/deepseek_v4_1_flash/decode_c2a_reuse.py [hunk 1]
```diff
@@ -6,41 +6,737 @@
...
+
+
+@pl.jit
+def decode_c2a_cache_test(
+    cache: pl.InOut[pl.Tensor[[CMP_BLOCKS_DYN, 128, 1, HEAD_DIM // 2], pl.UINT8]],
+    scales: pl.InOut[pl.Tensor[[CMP_BLOCKS_DYN, 128, 1, HEAD_DIM // 16], pl.FP8E4M3FN]],
+    output: pl.Out[pl.Tensor[[POOL_ROWS, HEAD_DIM], pl.BF16]],
+):
+    cache.bind_dynamic(0, CMP_BLOCKS_DYN)
+    output = decode_compressed(cache, scales, output)
```
[the remark this answers is not in the document]
Comment (indigo1973) [reply models/deepseek_v4_1_flash/decode_c2a_reuse.py]: 这是一个遗漏的中间测试过程的接口，已经清除。

**Comment by lwDavid** [comment]
I went through this because C2A full and C2A reuse publish into the same window
pool and read the same compressed pool, so I wanted to be sure the two would not
disagree once both are on main. Everything below was measured on 238 against
main at `7e22b2a`, not read off the diff. Offered as help, not as a gate.

## What I verified is correct

The reuse semantics are right, and I could not break them:

- `compressed_indices` is a pure input, there is no compressor and no indexer,
  and the compressed and index pools are never written.
- The packed cache ABI matches the merged one exactly. I exercised the nibble
  decode at `decode_c2a_reuse.py:297-309` over **all 256 byte values** against
  `quantization.dequantize_mxfp4_cache` — bit-exact, including nibble order
  (low nibble is logical element 2i), the E2M1 value table, and the group-16
  E4M3 scale.
- The Full-to-Reuse hand-off works. I fed your device kernel the compressed pool
  and Top-K row list that `official_reference_c2a` actually published for the
  same batch, and it reproduced that reference's output on an A5 die with zero
  outliers at your own tolerance, on two independent fixtures.
- `attention_tp.decode_tp_output_all_reduce` is imported unchanged, so the epoch
  protocol and the four `allow_early_resolve=False` waits are main's.
- Your own entry passes: `--tp 1 --tokens 16` and `--tp 4 --tokens 16 --epochs 4`
  both PASS on all three comparators.

So the operator is not broken. The two issues below are about it being validated
against a reference that #1241 replaced, which then pulled two changes into the
kernel itself.

## 1. `requantize_window` rewrites the whole window cache

`decode_c2a_reuse.py:171-203`, called at `:461`.

Measured with your own fixture (`build_values(32, seed=41, case='mixed')`): the
pool has 36,864 physical rows, 32 are mapped by a non-negative slot, and the
operator changes **15,812 unmapped rows — 646,139 payload bytes and 20,192 scale
bytes — while the decoded values are bit-for-bit unchanged (max |Δ| = 0.0)**.

The cause is that an MXFP8 round trip is not byte-idempotent: re-quantizing a
dequantized group whose largest magnitude does not sit at the top of its binade
re-derives a scale one step lower and moves the bytes without moving the values.
I measured this directly: one round trip over a 64x128x1x512 block changed
150,368 of 4,194,304 payload bytes at max value change 0.0.

Why it matters beyond the wasted ~9.4 MB of HBM traffic per layer per step: the
window pool is paged and shared across the DP group's active requests. A reuse
layer decoding one token for request 0 rewrites roughly 43% of every other
request's cached rows. `decode_c2a_full.py:1704-1722` makes "only rows named by
a non-negative slot may change" a checked byte contract, so this fails there
with `window_cache: unmapped cache bytes were modified`.

If this was added because a device run mismatched the golden, the fix that
landed on main is the opposite one: give the **reference** the row-scoped
publish (`official_reference_c2a` section b, `_write_rows(...)`), not give the
**device** a whole-pool rewrite.

## 2. `WOA_ACCUM_K = 316`

`decode_c2a_reuse.py:377`, used at `:381-411`.

This splits the `wo_a` reduction into 13 segments and rounds to BF16 after each,
to match cumulative BF16 rounding in the CPU torch reference. Measured for the
golden's exact shape on 238: `step=316` matches `torch.einsum` in BF16 on
16382/16384 elements, but matches a single FP32 accumulation on only
6486/16384.

Propagated through the operator, that rounding choice alone moves the output by
**rel-L2 0.016456, max row 0.018601, cosine 0.99986**.

The A5 kernel accumulates this GEMM in FP32 and casts once
(`decode_swa.grouped_output`: `pl.create_tensor(pl.FP32)` + `pl.matmul_acc` +
one `pl.cast`). `decode_c2a_full.py:1187-1198` pins the two CPU/device
disagreement points as named switchable constants, `BF16_GEMM = "fp32"` and
`PV_DTYPE = torch.bfloat16`, defaulting to what the device does and marked
"A/B only" — precisely so the device would not have to imitate a torch CPU
blocking artifact. As written, layers 3-7 / 9-13 / 15-19 would emit output 1.65%
worse than layers 2/8/14 compute from the same weights, and it would miss the
1% / cosine 0.9999 budget the full operator is held to.

Worth noting the constant is also fragile: oneDNN's K blocking depends on build,
thread count and shape, so 316 is not portable.

## 3. Two smaller ones

**Window quantizer floor.** `:219-228` floors the group amax at `448.0*2.0**-126`
plus a zero special case; `decode_swa.py:210` and the official transcription both
floor at `1e-4`. An exactly-zero published row then stores a different E8M0 code
than SWA and C2A full store for the same row. Your `--case` offers
mixed/window/long, so no fixture reaches it — the merged entry has `zero` and
`masked` cases for exactly this.

**`num_tokens` vs `pl.tensor.dim`.** `gather_cache:269`, `sparse_attention:333`,
`output_group_projection:386`, `make_projection:51`, `make_norm:115`, `rotate:146`
size their SPMD extents from the tensor height; only `publish_window:219` honours
the `num_tokens` scalar. With a padded `T_DYN` batch the rows past `num_tokens`
carry uninitialized indices, and any non-negative garbage becomes a physical
cache row and an out-of-bounds read of the paged pool. Main drives every stage
from the scalar. Your harness always sets `tokens == num_tokens`, so nothing
exercises it.

## 4. Most of this already exists on main

This is the one I would weigh most, because it is the difference between ~1130
lines and a couple of hundred. These all have validated equivalents:

| in this PR | already on main |
|---|---|
| `make_projection` (:42) | `decode_swa.make_projection` → `project_qa/qb/kv/ob` |
| `make_norm` (:110) | `decode_swa.make_norm` → `normalize_q/kv` |
| `rotate` (:141) | `decode_swa.make_rope` → `rotate_q/kv/output` |
| `publish_window` (:207) | `decode_swa.publish_window` |
| `gather_cache` (:248) | `decode_c2a_full.gather_sparse` |
| `sparse_attention` (:322) | `decode_c2a_full.attend_sparse` |
| `output_group_projection` (:386) | `decode_swa.grouped_output` |
| fixtures / comparators / specs | `decode_c2a_full.make_c2a_inputs`, `compare_output`, `compare_cache`, `compare_replicated`, `compare_per_rank`, `build_specs` |

I checked `gather_cache` and `sparse_attention` against their merged
counterparts and they agree bit-for-bit — they are correct, just not the copies
CI covers.

Concretely, a reuse partial is `c2a_full_partial` with the compressor, the
index-key publication and `index_select` deleted and `topk_indices` taken as an
argument. I did that for the prefill side in #1242: 546 lines including the
reference, the fixture and the harness, and it passed on device first try
because the leaf kernels were already validated. The same shape should collapse
this file to roughly the same size, and it is the convention `prefill_swa.py`
and `prefill_c2a_full.py` already follow — both import their leaf kernels from
the decode file rather than restating them.

## Suggested order

1. Drop `requantize_window` and publish only the mapped rows.
2. Drop `WOA_ACCUM_K` and accumulate `wo_a` in FP32.
3. Re-validate against `official_reference_c2a`-style comparators rather than
   `golden_compressed_attention`, with `zero` and `masked` in the case set.
4. Then cut the duplicated helpers over to imports.

Happy to be wrong on any of this — every number above is reproducible and I can
share the scripts. And if it is easier, I am glad to take the decode side onto
the shared pipeline myself and leave you the review; your call.

**Comment by lwDavid** [comment]
Re-reviewed against the rewrite. All four things I raised last time are genuinely
fixed, the operator agrees with the merged C2A full everywhere I could think to
check, and it passes on device. One line stands between this and mergeable.

Everything below was measured on 238 against the head commit, not read off the
diff.

## The one blocker: no `# ci: a5`

The header stops at `# ci: no-sim`. Since #1244 landed, that marker is the only
way an A5-only entry reaches the `a5-v41-flash` job. Combined with the split
`__main__` sentinel, the result is that no selector reaches this file at all —
and because the path already exists on main as a golden-smoke entry with a real
`__main__`, this is a **net loss** of coverage rather than a missing addition.
Simulated on a tree that is current main plus this file:

| version of the file | a2a3 / sim selects | a5 job selects |
|---|---|---|
| today's main (135-line golden smoke) | 1 | 0 |
| this PR | **0** | **0** |
| this PR + `# ci: a5` | 0 | **1** |

After merge the only jobs touching these 674 lines would be `pre-commit` and
`unit-tests`, and `unit-tests` imports nothing from it.

```python
 # ci: no-sim
+# ci: a5
```

Worth adding for a second reason specific to this design: with the marker, a
change to `decode_swa.py` or `decode_c2a_full.py` also pulls this entry into the
A5 job. That matters here more than for any other entry, because the whole file
is now built out of those shared kernels — verified, a `decode_swa.py` change
then selects all five entries including this one.

This is a rebase artifact, not an authoring mistake: the marker did not exist at
this branch's merge base.

## The four earlier items, re-checked

| item | status | measured |
|---|---|---|
| `requantize_window` rewrote the whole window pool | fixed — no local publish helper survives, only the shared `publish_window` | unmapped rows changed by the reference: 15,812 → **0**; device reports `untouched_exact=True` |
| `WOA_ACCUM_K = 316` segmented BF16 accumulation | fixed — shared `grouped_output`, one FP32 accumulator, one cast; reference uses `_bf16_grouped` | output rel-L2 **0.016456 → 0.0019**, cosine 0.99986 → 0.99999 |
| window quantizer floored at `448.0*2.0**-126` | fixed — no local quantizer at all; the only floor in the path is `1e-4` | `--case zero`: `window_cache_scale byte_mismatch=0/320` |
| stages sized from `pl.tensor.dim` instead of `num_tokens` | fixed — all 11 stages, the chunk loop and the TP reduction are scalar-driven; `pl.tensor.dim` only sizes scratch | padded run (40 tokens / 33 active) publishes exactly 33 rows and leaves the inactive tail at its init fill, with slot sentinels planted in the padding |
| ~900 duplicated leaf kernels | fixed — zero `pl.spmd` bodies in the file, all 14 device kernels imported | 1130 → 776 lines |

## Agreement with the merged C2A full

Checked from artifacts rather than by reading, since these two share a window
pool and a compressed pool:

- **Cache ABI** — compared the compiled parameter metadata of both entries. They
  agree exactly on every shared tensor: window cache E4M3 `[-1,128,1,512]` with
  group-32 E8M0 scales, compressed payload packed E2M1 as `uint8`
  `[-1,128,1,256]` with group-16 E4M3 scales, and `int32 [-1,512]` top-K.
- **Read-only discipline** — the compressed payload, its scales and the Top-K
  list compile to `ParamDirection.In` where full has them `InOut`, and there is
  no index-key or compressor-state parameter at all. That is the property I most
  wanted to be enforced at the ABI rather than in the body.
- **Reference convention** — `official_reference_c2a_reuse` returns
  `torch.equal`-true output and a byte-identical window cache against
  `official_reference_c2a` on all four fixture cases. It keeps FP32 `wo_a` and
  the BF16 PV cast, so it does not reintroduce the ~1.7% convention gap.
- **Call sites** — argument order, `num_tokens` threading and TaskId `deps` are
  line-for-line identical to `c2a_full_partial`; the only differences are the
  deleted compressor/indexer stages and one rename.
- **Group signature** — `pl.InOut` present on `window_cache`,
  `window_cache_scale` and `output`, and `pl.MX_B_NN` correctly dropped where
  the rank signature keeps it. That is the failure mode that silently validates
  nothing while printing PASS, so I checked it from the compiled directions.

## Device

Ran the entry myself on an A5 die:

```
[PRECISION] window_cache published_rows=32 byte_mismatch=0/16384 value_rel_l2=0 untouched_exact=True
[PRECISION] output rel_l2=0.00190599 cosine=0.99999356 max_abs=0.0078125 zero_rows_exact=True
[RUN] validate done (0.05s)
[RUN] PASS (30.02s)
```

Also green across `mixed` / `long` / `masked` / `zero`, a padded batch, capacity
(192 tokens x 32 requests), `--epochs 3`, save/replay, TP4 on four cards and
TP1/DP2 on two. The multi-epoch run is real rather than a no-op — device wall is
~3x the single-epoch one — and the single-pass golden is legitimate here because
reuse is idempotent, unlike full mode.

One thing to know rather than fix: `--case masked` compares an all-zero output
against an all-zero reference with zero published rows, so a green `masked` is
not evidence about the attention path. `zero` still exercises gather and
publication. The cases that actually validate the arithmetic are `mixed` and
`long`. The merged full entry behaves identically on the same device, so this is
the family's contract, not something introduced here.

## Non-blocking

1. `input_metadata` (:582) hardcodes `mode="decode"` and `capacity` is
   unconditionally `C.DECODE_MAX_TOKENS` (:704), while `run_c2a_reuse` takes a
   `mode`. Unreachable today. It is worth closing now anyway, because the
   obvious next step is a prefill sibling reusing this runner the way
   `prefill_c2a_full.py` reuses `run_c2a` — at which point the fixture builds a
   4-page window cache while the spec declares 7, and nothing cross-checks the
   two.
2. Same function re-derives the page counts with literal `128` and `2` instead
   of the config constants, making it a second source of truth alongside
   `make_c2a_inputs`.
3. `REUSE_CACHE_SLOTS` values are never read — `compare_owned_cache` hardcodes
   the slot name. Either use them or make it a tuple of names.
4. Function-local `import torch` in `input_metadata`, with the module already
   importing it at the top.
5. A validation report in a comment, with the environment, thresholds and the
   TP4/TP1 numbers, would help whoever reviews this next; right now the PR body
   is the commit message.

## Not issues

Two things I looked at and decided against raising, in case someone else spots
them: keeping the public `golden_decode_c2a_reuse` after its smoke caller went
away matches what every merged sibling does, and the degenerate `masked`
comparison above is inherited from the merged fixture rather than added here.

Happy to be wrong on any of this — every number is reproducible and I can share
the commands.

**Review by lwDavid — APPROVED** [review]
Approved.
